include/meta/embeddings/cooccurrence_counter.h

```diff
@@ -0,0 +1,225 @@
+{
+
+/**
+ * A (target, context) pair used as the key in a cooccurrence hash table.
+ */
+struct cooccurrence_key
+{
+    constexpr cooccurrence_key(uint64_t targ, uint64_t ctx)
+        : target{targ}, context{ctx}
+    {
+        // nothing
+    }
+
+    uint64_t target;
+    uint64_t context;
+};
+
+inline bool operator==(const cooccurrence_key& a, const cooccurrence_key& b)
+{
+    return std::tie(a.target, a.context) == std::tie(b.target, b.context);
+}
+
+inline bool operator<(const cooccurrence_key& a, const cooccurrence_key& b)
+{
+    return std::tie(a.target, a.context) < std::tie(b.target, b.context);
+}
+
+template <class OutputStream>
+uint64_t packed_write(OutputStream& os, const cooccurrence_key& key)
+{
+    auto bytes = io::packed::write(os, key.target);
+    return bytes + io::packed::write(os, key.context);
+}
+}
+
+namespace hashing
+{
+template <>
+struct key_traits<embeddings::cooccurrence_key>
+{
+    static constexpr bool inlineable = true;
+    constexpr static embeddings::cooccurrence_key sentinel()
+    {
+        return {key_traits<uint64_t>::sentinel(),
+                key_traits<uint64_t>::sentinel()};
+    }
+};
+
+template <>
+struct is_contiguously_hashable<embeddings::cooccurrence_key>
+{
+    const static constexpr bool value = true;
+};
+}
+
+namespace embeddings
+{
+
+/**
+ * A chunk of cooccurrence records on disk.
+ */
+struct cooccurrence_chunk
+{
+    cooccurrence_chunk(const std::string& file, uint64_t bytes)
+        : path{file}, size{bytes}
+    {
+        // nothing
+    }
+
+    std::string path;
+    uint64_t size;
+};
+
+inline bool operator<(const cooccurrence_chunk& a, const cooccurrence_chunk& b)
+{
+    // merge smaller chunks first
+    return a.size > b.size;
+}
+
+/**
+ * An iterator adhering to the ChunkIterator concept for multiway_merge
+ * support on in-memory cooccurrence data.
+ */
+class memory_cooccur_iterator
+{
+  public:
+    using map_type = hashing::probe_map<cooccurrence_key, double>;
+    using memory_chunk_type = map_type::storage_type::vector_type;
+    using count_type = std::pair<cooccurrence_key, double>;
+
+    memory_cooccur_iterator() = default;
+
+    memory_cooccur_iterator(memory_chunk_type&& items)
+        : items_{std::move(items)}, idx_{0}
+    {
+        // nothing
+    }
+
+    memory_cooccur_iterator& operator++()
+    {
+        ++idx_;
+        if (idx_ >= items_.size())
+        {
+            items_.clear();
+            idx_ = 0;
+        }
+
+        return *this;
+    }
+
+    cooccur_record operator*() const
+    {
+        const auto& item = items_[idx_];
+        return {item.first.target, item.first.context, item.second};
+    }
+
+    uint64_t total_bytes() const
+    {
+        return sizeof(count_type) * items_.size();
+    }
+
+    uint64_t bytes_read() const
+    {
+        return sizeof(count_type) * idx_;
+    }
+
+    bool operator==(const memory_cooccur_iterator& other) const
+    {
+        return items_.empty() && other.items_.empty();
```

Comment:
We only want `it1 == it2` when `it1` and `it2` are both the "end" iterator. This is only true when `items_` is empty.